homeassistant/components/hunterdouglas_powerview/cover.py

```diff
@@ -337,11 +469,131 @@ class PowerViewShade(PowerViewShadeBase):
         | CoverEntityFeature.SET_POSITION
     )
 
+
+class PowerViewShadeTDBU(PowerViewShade):
+    """Representation of a PowerView shade with top/down bottom/up capabilities."""
+
+    @property
+    def current_cover_position_primary(self):
+        """Return the current position of cover."""
+        return hd_position_to_hass(self.position_primary)
+
+    @property
+    def current_cover_position_secondary(self):
+        """Return the current position of cover."""
+        return hd_position_to_hass(self.position_secondary)
+
+    @property
+    def get_transition_steps(self):
+        """Return the steps to make a move."""
+        return (
+            self.current_cover_position_primary + self.current_cover_position_secondary
+        )
+
+    async def _async_move(self, target_hass_position):
+        """Move the shade to a position."""
+        # custom move command to prevent excessive refresh on tdbu
+        self._async_cancel_scheduled_transition_update()
+        self._forced_resync = async_call_later(
+            self.hass, RESYNC_DELAY, self._async_force_resync
+        )
+
+        await self._shade.move(self._set_shade_postion(target_hass_position))
+
+        self._async_cover_transition_complete(
+            self.current_cover_position, target_hass_position
+        )
+
+
+class PowerViewShadeTDBUBottom(PowerViewShadeTDBU):
+    """Representation of a top down bottom up powerview shade."""
+
+    def __init__(self, coordinator, device_info, room_name, shade, name):
+        """Initialize the shade."""
+        super().__init__(coordinator, device_info, room_name, shade, name)
+        self._attr_unique_id = f"{self._shade.id}_bottom"
+        self._attr_name = f"{self._shade_name} Bottom"
+
     @callback
-    def _async_process_updated_position_data(self, position_data):
-        """Process position data."""
-        if ATTR_POSITION1 in position_data:
-            self._current_hd_cover_position = int(position_data[ATTR_POSITION1])
+    def _clamp_cover_limit(self, target_hass_position):
+        """Dont allow a cover to go into an impossbile position."""
+        cover_top = self.current_cover_position_secondary
+        target_hass_position = min(target_hass_position, (100 - cover_top))
+        self.set_position_primary(hass_position_to_hd(target_hass_position))
```

Comment:
This function is doing more than clamping the value.  We shouldn't change the state in a clamp function.